apollo-router-core/src/spec/query.rs

```diff
@@ -595,85 +840,1527 @@ mod tests {
+        );
+
+        assert_format_response!(
+            schema,
+            "query { list { l2 } }",
+            json! {{
+                "list": {
+                    "l2": ["abc", "def"],
+                    "name": 1,
+                },
+            }},
+            None,
+            json! {{
+                "list": {
+                    "l2": ["abc", "def"],
+                },
+            }},
+        );
+
+        assert_format_response!(
+            schema,
+            "query { list { l3 } }",
+            json! {{
+                "list": {
+                    "l3": ["abc", 1, { "foo": "bar"}, ["aaa"], "def"],
+                    "name": 1,
+                },
+            }},
+            None,
+            json! {{
+                "list": {
+                    "l3": ["abc", null, null, null, "def"],
+                },
+            }},
+        );
+
+        assert_format_response!(
+            schema,
+            "query { list { l3 } }",
+            json! {{
+                "list": {
+                    "l3": 1,
+                },
+            }},
+            None,
+            json! {{
+                "list":null,
+            }},
+        );
+
+        assert_format_response!(
+            schema,
+            "query { list { l4 } }",
+            json! {{
+                "list": {
+                    "l4": ["abc", 1, { "foo": "bar"}, ["aaa"], "def"],
+                },
+            }},
+            None,
+            json! {{
+                "list": null,
+            }},
+        );
+
+        assert_format_response!(
+            schema,
+            "query { list { l4 } }",
+            json! {{
+                "list": {
+                    "l4": ["abc", "def"],
+                },
+            }},
+            None,
+            json! {{
+                "list": {
+                    "l4": ["abc", "def"],
+                },
+            }},
+        );
+
+        assert_format_response!(
+            schema,
+            "query { list { l4 } }",
+            json! {{
+                "list": {
+                    "l4": 1,
+                },
+            }},
+            None,
+            json! {{
+                "list": null,
+            }},
+        );
+    }
+
+    #[test]
+    fn filter_nested_object_errors() {
+        let schema = "type Query {
+            me: User
+        }
+
+        type User {
+            id: String!
+            name: String
+            reviews1: [Review]
+            reviews2: [Review!]
+            reviews3: [Review!]!
+        }
+        
+        type Review {
+            text1: String
+            text2: String!
+        }
+        ";
+
+        let query_review1_text1 = "query  { me { id reviews1 { text1 } } }";
+        assert_format_response!(
+            schema,
+            query_review1_text1,
+            json! {{
+                "me": {
+                    "id": "a",
+                    "name": 1,
+                    "reviews1": [ { } ],
```

Comment:
Is this a valid query? `{ }` Doesn't have any fields.

Reply:
Maybe empty selection sets are valid. Can't see anywhere in the spec that says they are not.

Reply:
Looks like selection sets cannot currently be empty: https://github.com/graphql/graphql-spec/issues/568

Reply:
I'm not sure about the meaning for that specific case. `reviews1` is an array, an empty object would have a different meaning than a null, right?

Reply:
I don't think it'll have a direct impact on the intent of the test. 
However the empty object is counted as a selection set, and I think these must have at least one field in them. If we add the validation that picks this up later it'll mean modifying all the tests. Maybe @abernix can confirm?

Reply:
that raises an interesting question: if a subgraph returns `{ }` instead of `{ text1: null }`, what should the router do? If it considers that the subgraph returned an invalid value, it should replace it entirely with `null`. If the router is more forgiving, it could replace with `{ text1: null }`

I think the router should be forgiving here (we can apply Postel's law for communication from the subgraph to the client)